src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs

```diff
@@ -0,0 +1,212 @@
+    {
+        throw new NotSupportedException("Attachment tombstones aren't supported by Queue ETL");
+    }
+
+    protected override IEnumerator<QueueItem> ConvertCountersEnumerator(DocumentsOperationContext context,
+        IEnumerator<CounterGroupDetail> counters, string collection)
+    {
+        throw new NotSupportedException("Counters aren't supported by Queue ETL");
+    }
+
+    protected override IEnumerator<QueueItem> ConvertTimeSeriesEnumerator(DocumentsOperationContext context,
+        IEnumerator<TimeSeriesSegmentEntry> timeSeries, string collection)
+    {
+        throw new NotSupportedException("Time series aren't supported by Queue ETL");
+    }
+
+    protected override IEnumerator<QueueItem> ConvertTimeSeriesDeletedRangeEnumerator(DocumentsOperationContext context,
+        IEnumerator<TimeSeriesDeletedRangeItem> timeSeries, string collection)
+    {
+        throw new NotSupportedException("Time series aren't supported by Queue ETL");
+    }
+
+    protected override EtlTransformer<QueueItem, QueueWithEvents, EtlStatsScope, EtlPerformanceOperation> GetTransformer(
+        DocumentsOperationContext context)
+    {
+        return new QueueDocumentTransformer(Transformation, Database, context, Configuration);
+    }
+
+    protected override int LoadInternal(IEnumerable<QueueWithEvents> items, DocumentsOperationContext context, EtlStatsScope scope)
+    {
+        int count = 0;
+
+        switch (Configuration.Connection.Provider)
+        {
+            case QueueProvider.Kafka:
+                count = ProcessKafka(items, Configuration, context);
+                break;
+            case QueueProvider.RabbitMq:
+                ProcessRabbitMq();
+                break;
+        }
+
+        return count;
+    }
+
+    protected override EtlStatsScope CreateScope(EtlRunStats stats)
+    {
+        return new EtlStatsScope(stats);
+    }
+
+    private int ProcessKafka(IEnumerable<QueueWithEvents> items, QueueEtlConfiguration queueEtlConfiguration,
+        DocumentsOperationContext context)
+    {
+        int count = 0;
+
+        using (var jsonSerializer = new BlittableEventBinaryFormatter(context))
+        {
+            var events = new List<KafkaMessageEvent>();
+
+            foreach (QueueWithEvents queueItem in items)
+            {
+                var topicName = queueItem.Name;
+
+                foreach (var insertItem in queueItem.Inserts)
+                {
+                    var eventMessage = new CloudEvent
+                    {
+                        Id = DefaultCloudEventId,
+                        DataContentType = "application/json",
+                        Type = string.IsNullOrWhiteSpace(queueItem.Type) ? DefaultType : queueItem.Type,
+                        Source = new Uri(string.IsNullOrWhiteSpace(queueItem.Source) ? DefaultSource : queueItem.Source),
+                        Data = insertItem.TransformationResult
+                    };
```

Comment:
Let's revise our approach when it comes to Id, Type ans Source. In particular according to the specs id + source should be unique.

1. ID - https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#id - I think that change vector would be suitable

2. Type - https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#type - looks like it's more up to a user's decision. I wonder if we shouldn't provide a way to let him provide it. e.g. `loadToUsers({ ... json message ... }. { type: 'some-type' })`


3. https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#source-1 - also user might be interested in setting it up but as the default I'd say that a node URL + database name + etl task name would be good candidate

Reply:
Discussed directly